crates/net/network/src/session/pending.rs

```diff
@@ -0,0 +1,161 @@
+                        direction,
+                        error: Some(PendingSessionHandshakeError::Eth(err.into())),
+                    };
+                }
+            };
+
+            // Check if we need to install extra protocols via multiplexing
+            let (conn, their_status) = if p2p_stream.shared_capabilities().len() == 1 {
+                let mut status_msg = handshake_info.status_msg;
+                status_msg.set_eth_version(eth_version);
+                let eth_unauthed = UnauthedEthStream::new(p2p_stream);
+                match eth_unauthed.handshake(status_msg, handshake_info.fork_filter.clone()).await {
+                    Ok((eth_stream, their_status)) => (eth_stream.into(), their_status),
+                    Err(err) => {
+                        return PendingSessionEvent::Disconnected {
+                            remote_addr: session_info.remote_addr,
+                            session_id: session_info.session_id,
+                            direction,
+                            error: Some(PendingSessionHandshakeError::Eth(err)),
+                        };
+                    }
+                }
+            } else {
+                let mut multiplex_stream = RlpxProtocolMultiplexer::new(p2p_stream);
+
+                for handler in extra_handlers.into_iter() {
+                    let cap = handler.protocol().cap;
+                    let remote_peer_id = their_hello.id;
+                    multiplex_stream
+                        .install_protocol(&cap, move |conn| {
+                            handler.into_connection(direction, remote_peer_id, conn)
+                        })
+                        .ok();
+                }
+
+                match multiplex_stream
+                    .into_eth_satellite_stream(
+                        handshake_info.status_msg,
+                        handshake_info.fork_filter,
+                    )
+                    .await
+                {
+                    Ok((multiplex_stream, their_status)) => (multiplex_stream.into(), their_status),
+                    Err(err) => {
+                        return PendingSessionEvent::Disconnected {
+                            remote_addr: session_info.remote_addr,
+                            session_id: session_info.session_id,
+                            direction,
+                            error: Some(PendingSessionHandshakeError::Eth(err)),
+                        };
+                    }
+                }
+            };
+
+            PendingSessionEvent::Established {
+                session_id: session_info.session_id,
+                remote_addr: session_info.remote_addr,
+                local_addr: session_info.local_addr,
+                peer_id: their_hello.id,
+                capabilities: Arc::new(Capabilities::from(their_hello.capabilities)),
+                status: Arc::new(their_status),
+                conn,
```

Comment:
this is the thing that's still problematic because this is still limited to the concrete connection type that operates on EthMessage

which I think we still want because we expect those variants, but conn here is of a concrete type so we either make this part of the ethhandler trait or we use a boxed stream as the type instead